cmd/olclient/stake.go

```diff
@@ -0,0 +1,140 @@
+	}
+}
+
+var stakeCmd = &cobra.Command{
+	Use:   "stake",
+	Short: "Perform stake to a validator",
+	RunE:  delegateStake,
+}
+
+var stakeArgs = &StakeArguments{}
+
+func setStakeArgs() {
+	// Transaction Parameters
+	stakeCmd.Flags().Int64Var(&stakeArgs.Amount, "amount", 0, "specify an amount")
+	stakeCmd.Flags().BytesHexVar(&stakeArgs.Address, "address", []byte{}, "address for account to pay the stake and fee")
+	stakeCmd.Flags().StringVar(&stakeArgs.Name, "name", "", "name for the validator, default to the node name in config.toml")
+	stakeCmd.Flags().StringVar(&stakeArgs.Password, "password", "", "password to access secure wallet")
+}
+
+func init() {
+	DelegationCmd.AddCommand(stakeCmd)
+	setStakeArgs()
+}
+
+// IssueRequest sends out a sendTx to all of the nodes in the chain
+func delegateStake(cmd *cobra.Command, args []string) error {
+	ctx := NewContext()
+	ctx.logger.Debug("Have Stake Request", "stakeArgs", stakeArgs)
+
+	rootPath, err := filepath.Abs(rootArgs.rootDir)
+	if err != nil {
+		return err
+	}
+
+	cfg := &config.Server{}
+
+	//Prompt for password
+	if len(stakeArgs.Password) == 0 {
+		stakeArgs.Password = PromptForPassword()
+	}
+
+	//Create new Wallet and User Address
+	wallet, err := accounts2.NewWalletKeyStore(keyStorePath)
+	if err != nil {
+		ctx.logger.Error("failed to create secure wallet", err)
+		return err
+	}
+
+	//Verify User Password
+	usrAddress := keys.Address(stakeArgs.Address)
+	authenticated, err := wallet.VerifyPassphrase(usrAddress, stakeArgs.Password)
+	if !authenticated {
+		ctx.logger.Error("authentication error", err)
+		return err
+	}
+
+	err = cfg.ReadFile(cfgPath(rootPath))
+	if err != nil {
+		return errors.Wrapf(err, "failed to read configuration file at at %s", cfgPath(rootPath))
+	}
+
+	// Create message
+	fullnode := ctx.clCtx.FullNodeClient()
+
+	out, err := fullnode.Stake(stakeArgs.ClientRequest())
+	if err != nil {
+		ctx.logger.Error("Error in applying ", err.Error())
+		return err
+	}
+
+	//Sign Transaction with secure wallet, then append signature to signedTx.
+	signedTx := &action.SignedTx{}
+	err = serialize.GetSerializer(serialize.NETWORK).Deserialize(out.RawTx, signedTx)
+	if err != nil {
+		return errors.New("error de-serializing signedTx")
+	}
+
+	if !wallet.Open(usrAddress, stakeArgs.Password) {
+		ctx.logger.Error("failed to open secure wallet")
+		return errors.New("failed to open secure wallet")
+	}
+
+	//Sign Raw "Send" Transaction Using Secure Wallet.
+	pub, signature, err := wallet.SignWithAddress(signedTx.RawTx.RawBytes(), usrAddress)
+	if err != nil {
+		ctx.logger.Error("error signing transaction", err)
+		return err
+	}
+
+	signedTx.Signatures = append([]action.Signature{{Signer: pub, Signed: signature}}, signedTx.Signatures...)
+
+	packet, err := serialize.GetSerializer(serialize.NETWORK).Serialize(signedTx)
+	if packet == nil || err != nil {
+		return errors.New("error serializing packet: " + err.Error())
+	}
+
+	result, err := ctx.clCtx.BroadcastTxCommit(packet)
```

Comment:
Can we use BroadcastTXSync instead ? Commit might lead to a timeout waiting for block to be included .

Reply:
https://godoc.org/github.com/tendermint/tendermint/rpc/core#BroadcastTxSync
Does not wait for DeliverTx result. But we want to be make sure that it was broadcasted.

Reply:
Agreed , but commit might not always be a success when using a VMs in a bigger network , we can use subscribe to the TX result via the web-socket if needed